File: bokehjs/src/lib/protocol/receiver.ts
import type {ID} from "../core/types"
import {isString} from "../core/util/types"
import type {Header} from "./message"
import {Message} from "./message"

export type Fragment = string | ArrayBufferLike

export class Receiver {
  message: Message<unknown> | null = null

  protected _header: Header | null = null
  protected _content: unknown = null
  protected _buffer_ids: ID[] = []
  protected _buffers: Map<ID, ArrayBuffer> = new Map()
  protected _current_consumer: (fragment: Fragment) => void = this._ENVELOPE

  consume(fragment: Fragment): void {
    try {
      this._current_consumer(fragment)
    } catch (error) {
      this._reset()
      throw error
    }
  }

  protected _reset(): void {
    this._header = null
    this._content = null
    this._buffer_ids = []
    this._buffers = new Map()
    this._current_consumer = this._ENVELOPE
  }

  _ENVELOPE(fragment: Fragment): void {
    this._assume_text(fragment)
    this.message = null
    const {header, content, buffers} = Message.decode(fragment)
    if (new Set(buffers).size != buffers.length) {
      throw new Error("Expected buffer ids to be unique")
    }

    if (buffers.length == 0) {
      this.message = new Message(header, content)
    } else {
      this._header = header
      this._content = content
      this._buffer_ids = buffers
      this._current_consumer = this._BUFFER_PAYLOAD
    }
  }

  _BUFFER_PAYLOAD(fragment: Fragment): void {
    this._assume_binary(fragment)
    const id = this._buffer_ids[this._buffers.size]
    this._buffers.set(id, fragment)
    if (this._buffers.size == this._buffer_ids.length) {
      this.message = new Message(this._header!, this._content, this._buffers)
      this._reset()
    }
  }

  private _assume_text(fragment: Fragment): asserts fragment is string {
    if (!isString(fragment)) {
      throw new Error("Expected text fragment but received binary fragment")
    }
  }

  private _assume_binary(fragment: Fragment): asserts fragment is ArrayBuffer {
    if (!(fragment instanceof ArrayBuffer)) {
      throw new Error("Expected binary fragment but received text fragment")
    }
  }
}

File: src/bokeh/protocol/receiver.py
#-----------------------------------------------------------------------------
# Copyright (c) Anaconda, Inc., and Bokeh Contributors.
# All rights reserved.
#
# The full license is in the file LICENSE.txt, distributed with this software.
#-----------------------------------------------------------------------------
'''Assemble a JSON envelope and its ordered binary payloads into a message.'''

from __future__ import annotations

# Standard library imports
from typing import Any, Callable

# Bokeh imports
from ..core.serialization import Buffer
from ..core.types import ID
from .exceptions import ProtocolError, ValidationError
from .message import Header, Message

__all__ = (
    'Receiver',
)

type Fragment = str | bytes

class Receiver:
    '''Receive one message envelope followed by its declared binary payloads.'''

    _current_consumer: Callable[[Fragment], Message[Any] | None]
    _header: Header | None
    _content: dict[str, Any] | None
    _buffer_ids: list[ID]
    _buffers: list[Buffer]

    def __init__(self) -> None:
        self._reset()

    def _reset(self) -> None:
        self._current_consumer = self._ENVELOPE
        self._header = None
        self._content = None
        self._buffer_ids = []
        self._buffers = []

    def consume(self, fragment: Fragment) -> Message[Any] | None:
        '''Consume an envelope or binary payload and return a completed message.'''
        try:
            return self._current_consumer(fragment)
        except ProtocolError:
            self._reset()
            raise

    def _ENVELOPE(self, fragment: Fragment) -> Message[Any] | None:
        header, content, buffer_ids = Message.decode(self._assume_text(fragment))
        if not buffer_ids:
            return Message(header, content)

        self._header = header
        self._content = content
        self._buffer_ids = buffer_ids
        self._current_consumer = self._BUFFER_PAYLOAD
        return None

    def _BUFFER_PAYLOAD(self, fragment: Fragment) -> Message[Any] | None:
        payload = self._assume_binary(fragment)
        buffer_id = self._buffer_ids[len(self._buffers)]
        self._buffers.append(Buffer(buffer_id, payload))

        if len(self._buffers) != len(self._buffer_ids):
            return None

        if self._header is None or self._content is None:
            raise ValidationError("buffer payload received without a message envelope")
        message = Message(self._header, self._content, self._buffers)
        self._reset()
        return message

    def _assume_text(self, fragment: Fragment) -> str:
        if not isinstance(fragment, str):
            raise ValidationError(f"expected text fragment but received binary fragment for {self._current_consumer.__name__}")
        return fragment

    def _assume_binary(self, fragment: Fragment) -> bytes:
        if not isinstance(fragment, bytes):
            raise ValidationError(f"expected binary fragment but received text fragment for {self._current_consumer.__name__}")
        return fragment
